cpp/examples/arrow/exec_plan_examples.cc

```diff
@@ -0,0 +1,1122 @@
+// under the License.
+
+#include <memory>
+#include <utility>
+
+#include "arrow/compute/api.h"
+#include "arrow/compute/api_scalar.h"
+#include "arrow/compute/api_vector.h"
+#include "arrow/compute/cast.h"
+#include "arrow/compute/exec/exec_plan.h"
+#include "arrow/compute/exec/ir_consumer.h"
+#include "arrow/compute/exec/test_util.h"
+
+#include <arrow/dataset/dataset.h>
+#include <arrow/dataset/file_parquet.h>
+#include "arrow/dataset/file_base.h"
+#include "arrow/dataset/plan.h"
+#include "arrow/dataset/scanner.h"
+#include "arrow/dataset/dataset_writer.h"
+
+#include "arrow/io/interfaces.h"
+#include "arrow/io/memory.h"
+#include "arrow/io/slow.h"
+#include "arrow/io/transform.h"
+
+#include <arrow/result.h>
+#include <arrow/status.h>
+#include <arrow/table.h>
+
+#include <arrow/ipc/api.h>
+
+#include <arrow/util/future.h>
+#include "arrow/util/range.h"
+#include "arrow/util/thread_pool.h"
+#include "arrow/util/vector.h"
+
+// Demonstrate various operators in Arrow Streaming Execution Engine
+
+#define ABORT_ON_FAILURE(expr)                     \
+  do {                                             \
+    arrow::Status status_ = (expr);                \
+    if (!status_.ok()) {                           \
+      std::cerr << status_.message() << std::endl; \
+      abort();                                     \
+    }                                              \
+  } while (0);
+
+#define CHECK_AND_RETURN(expr)                     \
+  do {                                             \
+    arrow::Status status_ = (expr);                \
+    if (!status_.ok()) {                           \
+      std::cerr << status_.message() << std::endl; \
+      return EXIT_FAILURE;                         \
+    } else {                                       \
+      return EXIT_SUCCESS;                         \
+    }                                              \
+  } while (0);
+
+#define CHECK_AND_CONTINUE(expr)                   \
+  do {                                             \
+    arrow::Status status_ = (expr);                \
+    if (!status_.ok()) {                           \
+      std::cerr << status_.message() << std::endl; \
+      return EXIT_FAILURE;                         \
+    }                                              \
+  } while (0);
+
+#define SEP_STR "******"
+
+#define PRINT_BLOCK(msg)                                                     \
+  std::cout << "" << std::endl;                                              \
+  std::cout << "\t" << SEP_STR << " " << msg << " " << SEP_STR << std::endl; \
+  std::cout << "" << std::endl;
+
+#define PRINT_LINE(msg) std::cout << msg << std::endl;
+
+namespace cp = ::arrow::compute;
+
+std::shared_ptr<arrow::Array> GetArrayFromJSON(
+    const std::shared_ptr<arrow::DataType>& type, arrow::util::string_view json) {
+  std::shared_ptr<arrow::Array> out;
+
+  ABORT_ON_FAILURE(arrow::ipc::internal::json::ArrayFromJSON(type, json, &out));
```

Comment:
I'm not sure we want to use the internal JSON format, both because it's not documented and because its format differs from the JSONL format the actual JSON reader uses. Can we use either the actual JSON reader or the CSV reader?

Reply:
Yes, we can do that. Should we hard code some values or use a real world dataset for this?

Reply:
I think we can keep hardcoding values just to keep it self-contained.

Reply:
Fixed with a csv based reader using in-memory string data.